src/main/java/edu/bu/metcs673/trackr/service/impl/UserServiceImpl.java

```diff
@@ -25,23 +32,55 @@ public class UserServiceImpl implements UserService {
 	@Autowired
 	private UserRepository userRepository;
 
-	@Override
-	public List<User> retrieveAllUsers() {
-		// TODO Auto-generated method stub
-		return null;
-	}
-
 	@Override
 	public User findUserById(long id) {
 		log.info("Searching for record id: " + id);
-		User user = userRepository.findById(id);
+		User user = userRepository.findById(id).get();
 		return user;
 	}
 
 	@Override
-	public void createUser(String username, String password, String email) {
-		// TODO Auto-generated method stub
+	public ResponseEntity<GenericApiResponse> createUser(User userInput) {
+		// validate parameters
+		validateParameters(userInput);
+
+		// TODO: hash value for password field to increase security
+
+		// saves new User record in DB
+		userRepository.save(userInput);
+		
+		// return response entity with a success response
+		return new ResponseEntity<GenericApiResponse>(
+				GenericApiResponse.successResponse(CommonConstants.CREATE_USER_SUCCESS), HttpStatus.ACCEPTED);
+	}
+
+	/**
+	 * Validates the parameters of the User (username, password, email). Checks that
+	 * all have value, that username is unique, and that email has a valid format.
+	 * 
+	 * @param userInput
+	 */
+	public void validateParameters(User userInput) {
+
+		String username = userInput.getUsername();
+		String password = userInput.getPassword();
+		String email = userInput.getEmail();
+
+		// since all of these fields are required, check that they have some value
+		if (StringUtils.isBlank(username) || StringUtils.isBlank(password) || StringUtils.isBlank(email)) {
+			throw new TrackrInputValidationException(CommonConstants.NULL_PARAMETERS);
+		}
+
+		// check that username is unique
+		if (userRepository.existsByUsername(username)) {
+			throw new TrackrInputValidationException(CommonConstants.DUPLICATE_USERNAME);
+		}
 
+		// check that email has proper format (ex. myEmail@email.com)
+		String emailRegex = "^[a-zA-Z0-9_!#$%&'*+/=?`{|}~^.-]+@[a-zA-Z0-9.-]+$";
```

Comment:
I have found this on the web and there is one section about Apache Common modules for email validation. Let me know what you think instead of maintaining our own regex.
https://www.baeldung.com/java-email-validation-regex

Reply:
So I looked into this, before pushing and the Apache Commons EmailValidator was showing as deprecated. Maybe we could just move this regex to an `@Email` annotation on the User object?

Reply:
That sounds good to me.